src/psm/src/ir_solver.cpp

```diff
@@ -654,81 +653,120 @@ void IRSolver::createGmatNodes(const vector<dbSBox*>& power_wires,
+{
+  for (auto curWire : power_wires) {
+    // For a stripe we create nodes at the ends of the stripes and at a fixed
+    // frequency in the lowermost layer.
+    if (curWire->isVia())
+     continue;
+    dbTechLayer* wire_layer = curWire->getTechLayer();
+    const int l = wire_layer->getRoutingLevel();
+    dbTechLayerDir::Value layer_dir = wire_layer->getDirection();
+    if (l == bottom_layer_) {
+      layer_dir = dbTechLayerDir::Value::HORIZONTAL;
+    }
+    int x_loc1, x_loc2, y_loc1, y_loc2;
+    if (layer_dir == dbTechLayerDir::Value::HORIZONTAL) {
+      y_loc1 = (curWire->yMin() + curWire->yMax()) / 2;
+      y_loc2 = y_loc1;
+      x_loc1 = curWire->xMin();
+      x_loc2 = curWire->xMax();
     } else {
-      dbTechLayer* wire_layer = curWire->getTechLayer();
-      const int l = wire_layer->getRoutingLevel();
-      dbTechLayerDir::Value layer_dir = wire_layer->getDirection();
-      if (l == bottom_layer_) {
-        layer_dir = dbTechLayerDir::Value::HORIZONTAL;
-      }
-      int x_loc1, x_loc2, y_loc1, y_loc2;
+      x_loc1 = (curWire->xMin() + curWire->xMax()) / 2;
+      x_loc2 = x_loc1;
+      y_loc1 = curWire->yMin();
+      y_loc2 = curWire->yMax();
+    }
+    // For all layers we create the end nodes
+    Gmat_->setNode({x_loc1, y_loc1}, l);
+    Gmat_->setNode({x_loc2, y_loc2}, l);
+    // Special condition: if the stripe ovelaps a macro ensure a node is
+    // created
+    for (const auto& macro : macros) {
       if (layer_dir == dbTechLayerDir::Value::HORIZONTAL) {
-        y_loc1 = (curWire->yMin() + curWire->yMax()) / 2;
-        y_loc2 = y_loc1;
-        x_loc1 = curWire->xMin();
-        x_loc2 = curWire->xMax();
+        // y range is withing the marco (min, max)
+        if (y_loc1 >= macro.yMin() && y_loc1 <= macro.yMax()) {
+          // Both x values outside the macro
+          // (Values inside will already have a node at endpoints)
+          if (x_loc1 < macro.xMin() && x_loc2 > macro.xMax()) {
+            const int x = (macro.xMin() + macro.xMax()) / 2;
+            Gmat_->setNode({x, y_loc1}, l);
+          }
+        }
       } else {
-        x_loc1 = (curWire->xMin() + curWire->xMax()) / 2;
-        x_loc2 = x_loc1;
-        y_loc1 = curWire->yMin();
-        y_loc2 = curWire->yMax();
+        if (x_loc1 >= macro.xMin() && x_loc1 <= macro.xMax()) {
+          if (y_loc1 < macro.yMin() && y_loc2 > macro.yMax()) {
+            const int y = (macro.yMin() + macro.yMax()) / 2;
+            Gmat_->setNode({x_loc1, y}, l);
+          }
+        }
       }
-      // special case for bottom layers we design a dense grid at a fixed
-      // frequency
-      if (l == bottom_layer_) {
+    }
+    if (l != bottom_layer_)
+      continue;
+
+    // special case for bottom layers we design a dense grid at a fixed
+    // frequency
+    auto node_map = Gmat_->getNodes(l,layer_dir,x_loc1,x_loc2,y_loc1,y_loc2);
+    pair<pair<int, int>, Node*> node_prev;
```

Comment:
Just use a tuple or better yet a struct

Reply:
If this is with respect to line 720, I'm using a pair of pairs as this is the return type as I iterate through the "node_map" in the subsequent for loop. The format for node_map is used in other parts of the code so it's a bigger change.